machines: Use libosinfo APIs instead of calling its tools

fidencio opened the pull request:
Instead of calling libosinfo tools, let's rely on libosinfo APIs in
order to avoid doing a lot of parsing to match the detected OS from a
media or a tree.

In order to do so, the code in libvirt-common.js has been expanded to
also store the (unique) OS ID and, when detecting some media or tree,
then just use this ID when filtering the OS list in autodetectOS().

With the changes, two new python scripts have been added:
- autoDetectOS.py, which does the media/tree detection and returns the
  OS ID;
- getOSList.py, which gets exactly the same information (and in the same
  format) returned by osinfo-query, plus the OS ID.

Signed-off-by: Fabiano Fidêncio <fidencio@redhat.com>

KKoukiou commented on pkg/machines/components/create-vm-dialog/autoDetectOS.py:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/python
```
python3

KKoukiou replied:
Does it actually work wity python2 this script as well?

fidencio replied:
@KKoukiou, yes, both scripts work with python2 and python3

` ` `
fidencio@laerte ~/src/upstream/cockpit $ python2.7 pkg/machines/components/create-vm-dialog/autoDetectOS.py https://dl.fedoraproject.org/pub/fedora/linux/releases/29/Workstation/x86_64/os/
http://fedoraproject.org/fedora/29
fidencio@laerte ~/src/upstream/cockpit $ python2.7 pkg/machines/components/create-vm-dialog/autoDetectOS.py https://dl.fedoraproject.org/pub/fedora/linux/releases/29/Workstation/x86_64/os/images/boot.iso
http://fedoraproject.org/fedora/29
fidencio@laerte ~/src/upstream/cockpit $ python3.7 pkg/machines/components/create-vm-dialog/autoDetectOS.py https://dl.fedoraproject.org/pub/fedora/linux/releases/29/Workstation/x86_64/os/
http://fedoraproject.org/fedora/29
fidencio@laerte ~/src/upstream/cockpit $ python3.7 pkg/machines/components/create-vm-dialog/autoDetectOS.py https://dl.fedoraproject.org/pub/fedora/linux/releases/29/Workstation/x86_64/os/images/boot.iso
http://fedoraproject.org/fedora/29
fidencio@laerte ~/src/upstream/cockpit $ python2.7 pkg/machines/getOSList.py | wc -l
570
fidencio@laerte ~/src/upstream/cockpit $ python3.7 pkg/machines/getOSList.py | wc -l
570
```

KKoukiou commented on pkg/machines/getOSList.py:
```diff
@@ -0,0 +1,55 @@
+
+    osId = os.get_id()
+    if not osId:
+        osId = "";
```
Please don't use semicolons in the python code.

KKoukiou replied:
And pls fix all occurances of semicols below.

fidencio replied:
Ouch! That's my brain playing tricks with me due to context switch between js and py. :-)

KKoukiou commented on pkg/machines/getOSList.py:
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/python
```
python3

KKoukiou commented:
@fidencio And please remove the signed of line, we don't use it in cockpit ;)

fidencio commented:
Okay, I've addressed all the comments but the ones related to using python3 in the shebang, as the scripts also work with python2.

Please, let me know what would be the preference about that and I can update the patch without any issue.

Thanks for the quite fast review!

marusak commented:
> Okay, I've addressed all the comments but the ones related to using python3 in the shebang, as the scripts also work with python2

https://docs.fedoraproject.org/en-US/packaging-guidelines/Python/#_multiple_python_runtimes
> Packages in Fedora MUST NOT use /usr/bin/python

Since we ship to Fedora, maybe we should follow up the guidelines. Just wondering, do we (in Cockpit) care about python2 support? I guess that only `rhel-7` does not support python3, but we are not gonna backport this, right?

fidencio commented:
Makes sense.

Following "Packages in Fedora MUST NOT use /usr/bin/python. Instead packages for Python 3 MUST use /usr/bin/python3 (even if upstream supports both Python 2 and 3)" the patch has been updated.

martinpitt commented:
This will most likely be backported to RHEL/CentOS 7, and CI tests on these either way. So we do care about Python scripts being bilingual. This correctly uses `python.spawn()`, which will use python 3 or 2 depending on what's available. The shebangs should be python3 mostly just to pacify static scanners.

martinpitt commented on pkg/machines/components/create-vm-dialog/autoDetectOS.py:
```diff
@@ -0,0 +1,25 @@
+
+import gi
+gi.require_version('Libosinfo', '1.0')
+from gi.repository import Libosinfo
```
This requires adding a dependency (Recommends) to python(3)-gobject-base, plus the corresponding package that actually ships the Typelib; this needs to happen in tools/cockpit.spec and tools/debian/control.

Does that typelib exist on RHEL 7? Supposedly on these OSes, this should just fail and autodetection will simply be unsupported (which is fine).

fidencio replied:
Yep, it's part of RHEL-7. I've updated the deps according to your review.

martinpitt commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -134,12 +134,7 @@ class CreateVM extends React.Component {
-                                osName = osName.replace(/ Server$/, "");
-                                osName = osName.replace(/ Workstation$/, "");
-                                const osEntry = this.props.osInfoList.filter(osEntry => osEntry.name == osName);
+                                const osEntry = this.props.osInfoList.filter(osEntry => osEntry.id == res.trim());
```
Please don't repeatedly call trim() -- do it once outside the loop.

fidencio replied:
I've also asked on IRC, but let me ask here so I can easily get back to this if needed.

Calling trim() in the value returned by autodetectOS() is something that's actually bothering me. Ideally we should trim() the return in the autodetectOS() function, not on its callers.

However, my javascript knowledge is quite limited and I didn't find an easy way to do so.
If you have any suggestion, please, let me know.

martinpitt replied:
That's a bit tricky indeed, as `autodetectOS()` really just directly returns a promise from cockpit.spawn(). A cleaner way is to use [promise chaining](https://javascript.info/promise-chaining), i. e. define `autodetectOS()` like this:

    return cockpit.spawn(...).then(out => out.trim());

martinpitt commented on pkg/machines/getOSList.py:
```diff
@@ -0,0 +1,55 @@
+
+oses = db.get_os_list()
+for i in range(oses.get_length()):
+    os = oses.get_nth(i)
```
This looks very unpythonic. You can't just iterate more naturally over `for os in oses:`?

fidencio replied:
That's a gobject introspection limitation, unfortunately.

fidencio replied:
Okay, I've worked this around in the same way done by virt-install.

martinpitt replied:
If it's a GI limitation, it's fine to keep. Still better than the large overhead of defining the helper class, IMHO.

fidencio replied:
Right, I'll move it back to the way it was.

For the future, in case Cockout starts relying on Libosinfo for resources/devices, I do believe that the helper class may become handy. However, for now, let me just get back to the original proposal. :-)

martinpitt commented on pkg/machines/getOSList.py:
```diff
@@ -0,0 +1,55 @@
+
+    osId = os.get_id()
+    if not osId:
+        osId = ""
```
Please use the pattern `varname = call() or ""` to avoid having to repeat this pattern so often.

martinpitt commented on pkg/machines/getOSList.py:
```diff
@@ -0,0 +1,55 @@
+
+    print("%s|%s|%s|%s|%s|%s|%s|%s|%s" %
+          (osId, osShortId, osName, osVersion, osFamily, osVendor,
+           osReleaseDate, osEOLDate, osCodename))
```
Wouldn't it make more sense to print JSON here, so that it's less error prone to read from the JS side?

fidencio replied:
It would. But, sincerely. I'd really prefer to **not** do this (and then having to change the code dealing with the result) in this patch.

This change could be done together with more improvements (as relying on Libosinfo for recommended resources/devices/machine type) later on, if you don't mind.

martinpitt replied:
Sure, it was just a suggestion, not a veto.

martinpitt commented on tools/cockpit.spec:
```diff
@@ -709,6 +709,11 @@ Requires: libvirt-client
 Recommends: virt-install
+Recommends: libosinfo-devel
+%if 0%{?rhel} == 7
+Recommends: python-gobject-base
```
RHEL/CentOS 7 don't yet support "Recommends". Also, this addition is in the `%if 0%{?fedora} || 0%{?rhel} >= 8` section, so just simply add the Recommends: --  Let's not force these on RHEL 7, it's ok if auto-detection does not work there yet.

-devel is not necessary, the typelib is in libosinfo already (as it should be).

fidencio replied:
Mind that listing the OSes in the vm-create dialog would also be broken in case the deps are not installed. With that in mind, wouldn't be better to have a Requires: entry for el7?

martinpitt commented on pkg/machines/getOSList.py:
```diff
@@ -0,0 +1,47 @@
+import sys
+
+
+class _OsinfoIter:
```
Too much overhead IMHO, the simple iteration was fine. I just wanted to know why it was done that way.

martinpitt commented on tools/debian/control:
```diff
@@ -104,9 +104,8 @@ Architecture: all
-         libosinfo-bin
-Recommends: virtinst, libvirt-dbus
+         libvirt-clients | libvirt-bin
+Recommends: virtinst, libvirt-dbus, python3-gi, libosinfo-1.0-dev
```
Likewise, no need for the -dev package; gir1.2-libosinfo-1.0 is enough (for the typelib).

martinpitt commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -134,12 +134,8 @@ class CreateVM extends React.Component {
-                                osName = osName.replace(/ Server$/, "");
-                                osName = osName.replace(/ Workstation$/, "");
-                                const osEntry = this.props.osInfoList.filter(osEntry => osEntry.name == osName);
+                                let osId = res.trim();
```
Let's rather fix autodetectOS() to trim the result by itself, see above.

martinpitt commented:
For the record: Our rhel7 images do have python-gobject-base and libosinfo already, they are a dependency of virt-install.
